### ISSUE TITLE
[Docs]: Bad Ceph advice

### ISSUE THREAD

**Issue by Ornias1993** [cloudnative-pg/cloudnative-pg#4355]
### Is there an existing issue already for your request/idea?

- [X] I have searched for an existing issue, and could not find anything. I believe this is a new documentation enhancement to be evaluated.

### What problem in the existing documentation this issue aims to solve?

The manual states the following:

> In these cases it makes sense to define the storage class used by the Postgres clusters to be defined as 1 replica. By having additional replicas defined in the storage solution like Longhorn and Ceph you might incur in the issue known as write amplification, unnecessarily increasing disk I/O and space used.

For Ceph, this is incredibly bad advice.
- 1 Replica, means that if 1 of **any** disks or nodes fails, the storage can be broken.
- 1 Replica also does **not** assure said replica's are on seperate disks or even nodes.

So simply put:
This advice might take down the complete postgresql cluster once a single(!) disk goes down on ceph

### Describe what additions need to be done to the documentation

On ceph a minimum adviced setup would be 2 replicas, preferably spread over different nodes.

### Describe what pages need to change in the documentation, if any

https://cloudnative-pg.io/documentation/1.20/storage/

### Backport?

Yes

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by gbartolini** [comment]
This is the whole purpose of a shared nothing architecture.

**Comment by Ornias1993** [comment]
> This is the whole purpose of a shared nothing architecture. See https://www.cncf.io/blog/2023/09/29/recommended-architectures-for-postgresql-in-kubernetes/ for details.

Whats your point really?
Let me clearify: **I KNOW** that local-storage is adviced. Thats also what I would advice anyone using CNPG.

The problem is your advice when using CEPH specifically, goes against **ANY** guidelines for CEPH in existence anywhere.
Well, it might work okey-ish for single-pod CNPG setups, but it absolutely should not be used on multi-pod setups **in any way**.
And with "work okey-ish" that means: Is potentially death in the water with a single disk-failure.

Simply put:
The advice is so inherently bad, that it's obviously writhen by someone with absolutely zero experience with CEPH.

**Comment by rouke-broersma** [comment]
@Ornias1993 I think their point is that you would spread the cnpg workload around in a shared-nothing way even if you use block storage like longhorn or ceph. You would spread your block storage around fault domains and you would spread your database replicas around fault domains. If a disk fails in fault domain 1, this should not affect database replica 2 which is in fault domain 2. Sure replica 1 will fail, but that replica will come back when you fix the storage issue affecting fault domain 1. You could spread the volumes for replica 1 around fault domain 1 and 2, and then replica 1 could remain online while fault domain 1 has a storage issue. However this is no longer a shared nothing architecture and as mentioned in the docs incurs a in this case unnecessary penalty. Instead you should simply rely on your replicas to be available in the other fault domains.

**Comment by Ornias1993** [comment]
@rouke-broersma 

I get their point, it's just wrong. Because they don't understand ceph.

2+ CNPG replica's on a CEPH pool with replica 1 set:
Does **NOT** guarantee both replicas are on seperate disks within the CEPH pool

**Comment by rouke-broersma** [comment]
@Ornias1993 

> No it won't "come back", replica 1 in ceph: if the disk with these placement groups is done-for/replaced, the data is gone.

The data is in replica 2, which is the point. Simply rebuild the replica from the cluster.

Look I'm gonna be honest I have not yet used Ceph in production, however I know Kubernetes PV is topology aware and I know CSI is topology aware.

I did a quick search, and found that Rook Ceph (Kubernetes CSI for Ceph) is also optionally topology aware. So it seems to me like this is simply a supported scenario in Rook Ceph.

See: https://rook.io/docs/rook/latest-release/CRDs/Cluster/external-cluster/topology-for-external-mode/?h=topology

So tbh yes it seems like this exact scenario is available in Ceph.

**Comment by Ornias1993** [comment]
@rouke-broersma 

> Look I'm gonna be honest I have not yet used Ceph in production

I'm going to be harsh here: This for sure is not your area of experise, which is fine.
But then leave it to people that have some more expertise than you do.

> however I know Kubernetes PV is topology aware and I know CSI is topology aware.

This has nothing to do with PV, PVC or CSI. 
This has to do with the basic fundementals of CEPH.

> I did a quick search

These are **EVERY** high level, advanced tricks.
That's precisely what I meant with "So the advice would only apply to a VERY specific niche setup design of CEPH."

Setup is basically the "niche-of-niches" of CEPH deployments, not normally done, not normally even adviced unless you've a dedicated storage team on-site with significant expertise in both CEPH and Kubernetes.

It also is limited to "external ceph cluster", according to those docs.

---

What you link does not mean "Just use replica 1".
The only thing I'm saying is that you do not "just use replica 1".
Thats simply not the full story and is **inherently** not data safe.

I never said that Replica 1 doesn't have a place, I never said that CEPH cannot work.
I said that the oneliner to "use replica 1" is bad and destructive advice.

That's just CEPH 101, the absolute bare-minimum basics.

**Comment by Ornias1993** [comment]
Issue is muted for me now, if it's not clear why its bad, the maintainers should just contact a ceph expert to explain it to them (again).

**Comment by Ornias1993** [comment]
Seeing the responses, i've decided not to put any effort into this.
Closing, a shame really.
